torchtune/models/llama2/_tokenizer.py

```diff
@@ -0,0 +1,141 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Tuple
+
+from torchtune.data import Message
+from torchtune.modules.tokenizers import (
+    ModelTokenizer,
+    SentencePieceBaseTokenizer,
+    tokenize_messages_no_special_tokens,
+)
+
+WHITESPACE_CHARS = [" ", "\n", "\t", "\r", "\v"]
+
+
+class Llama2Tokenizer(ModelTokenizer):
+    """
+    Llama2's implementation of the SentencePiece tokenizer. Llama2Tokenizer does
+    not include any additional special tokens. The prompt template described in
+    https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-2/ describes
+    [INST][/INST] and <<SYS>><</SYS>> as special tokens but these are not registered
+    as unique ids and are tokenized as normal text. When using this tokenizer on the
+    pre-trained model for inference, it is strongly encouraged to apply the
+    :class:`~torchtune.data.Llama2ChatFormat` to your data beforehand to add the
+    [INST] and <<SYS>> for optimal performance. For fine-tuning, this is not required.
+    For more details, see https://pytorch.org/torchtune/main/tutorials/chat.html#tokenizing-prompt-templates-special-tokens.
+
+    Args:
+        path (str): Path to pretrained SentencePiece tokenizer file.
+
+    Example:
+        # Accepts only non-batched input for now
+        >>> tokenizer = Llama2Tokenizer("/path/to/spm_model")
+        >>> tokenized_text = tokenizer.encode("Hello world!", add_bos=True, add_eos=True)
+        >>> print(tokenized_text)
+        [1, 31587, 29644, 102, 2]
+    """
+
+    def __init__(
+        self,
+        path: str,
+    ):
+        self._spm_model = SentencePieceBaseTokenizer(path)
+
+        # Original tokenizer has no pad_id, which causes indexing errors when batch training
+        self._spm_model.pad_id = 0
+
+        # This is used in tokenize_messages: if the tokenizer does not
+        # encode whitespace, then we can more easily split strings
+        # on whitespace characters and encode them separately.
+        self.encodes_whitespace = any(
+            [self._spm_model.encode(c) for c in WHITESPACE_CHARS]
+        )
```

Comment:
I don't like this being in the model-specific tokenizers. It's very much a hack to get `encode(a) + encode(b) = encode(a+b)` to hold for sentencepiece. But if I am a user copy-pasting tokenizers I am gonna have no idea what the hell is going on here